File: Technology Platforms (Java, .NET)/Java (multithreading, recursive class, socket server, database, unit testing)/src/main/java/org/example/SharedResults.java
package org.example;

import java.io.Serializable;

public class SharedResults implements Serializable {
    private double sum = 0;
    private double sumSquares = 0;
    private int count = 0;

    public synchronized void addPartialResult(double sum, double sumSquares, int count) {
        this.sum += sum;
        this.sumSquares += sumSquares;
        this.count += count;
    }

    public synchronized double getMean() {
        return sum / count;
    }

    public synchronized double getVariance() {
        return (sumSquares - (sum * sum) / count) / count;
    }

    public synchronized void printStatistics() {
        System.out.println("\nStatystyki pośrednie:");
        System.out.println("Łączna suma:        " + sum);
        System.out.println("Suma kwadratów:     " + sumSquares);
        System.out.println("Liczba elementów:   " + count);
        System.out.println("Średnia:            " + getMean());
        System.out.println("Wariancja:          " + getVariance());
        System.out.println("Odchylenie std:     " + Math.sqrt(getVariance()));
        System.out.println("\n");
    }
}

File: Technology Platforms (Java, .NET)/.NET (recursive class, LINQ, XML, and XPath, datagrid, multithreading)/SharedResults.cs
using System;

namespace WpfApp
{
    public class SharedResults
    {
        private double sum = 0;
        private double sumSquares = 0;
        private int count = 0;
        private readonly object lockObj = new object();
        private readonly int totalCount;

        public event Action<double> ProgressChanged;

        public SharedResults()
        {
            this.totalCount = totalCount;
        }

        public SharedResults(int totalCount)
        {
            this.totalCount = totalCount;
        }

        public void AddPartialResult(double sum, double sumSquares, int count)
        {
            lock (lockObj)
            {
                this.sum += sum;
                this.sumSquares += sumSquares;
                this.count += count;

                if (this.count % 1000 == 0)
                {
                    double progress = ((double)this.count / totalCount) * 50;
                    ProgressChanged?.Invoke(progress);
                }
            }
        }

        public double GetMean()
        {
            lock (lockObj)
            {
                return sum / count;
            }
        }

        public double GetVariance()
        {
            lock (lockObj)
            {
                return (sumSquares - (sum * sum) / count) / count;
            }
        }

        public void PrintStatistics()
        {
            lock (lockObj)
            {
                Console.WriteLine("\nStatystyki pośrednie:");
                Console.WriteLine("Łączna suma:        " + sum);
                Console.WriteLine("Suma kwadratów:     " + sumSquares);
                Console.WriteLine("Liczba elementów:   " + count);
                Console.WriteLine("Średnia:            " + GetMean());
                Console.WriteLine("Wariancja:          " + GetVariance());
                Console.WriteLine("Odchylenie std:     " + Math.Sqrt(GetVariance()));
            }
        }
    }
}
